### ISSUE TITLE
Convenience API to rerun failed promotion

### ISSUE THREAD

**Issue by jessesuen** [akuity/kargo#3159]
# Proposed Feature

A promotion might fail for an intermittent issue (e.g. in my case it was a timeout to github). A rerun of the promotion resolved it. But in order to rerun it, I had to navigate the UI, select the same freight, select the Stage, click promote, etc... It would be easier to make this one click.

# Motivation

It will be common for promotions to fail for ephemeral reasons and we should make it easy to retry it.

# Suggested Implementation

* When looking at a failed promotion, UI should display a rerun button
* When rerun, a new promotion would be created using the same steps/vars/freight etc.. of the failed promo.

NOTE: at first, I thought this would just be a UI feature. But then I thought we may need to do this at the API level. Reasons:
1. We may want to preserve the promotion steps of the failed promotion, in case the promotionTemplate had changed in the Stage with different steps. The logic of this would be too involved for the UI, so the API server would need to do it.
2. CLI might want to do this too
3. We might considered the ability to rerun a failed promotion, a different RBAC permission than `promote`. Use case: end user may not be given `promote` privileges since promotions are automatically created via autoPromote. But when the promotion fails for an ephemeral issue, users should be able to rerun it.

**Comment by krancour** [comment]
💯 on the idea. Also agree on adding an API endpoint for it.

> We may want to preserve the promotion steps of the failed promotion, in case the promotionTemplate had changed in the Stage with different steps.

I feel iffy about this one aspect. Supposing the failure _weren't_ for ephemeral reasons, but due to a flaw in the process, I can easily imagine a user updating the Stage's promotionTemplate and then retrying the promotion. In this case, you'd _want_ to pick up the changes.

Supposing the failure _were_ for ephemeral reasons and promotionTemplate had been updated for entirely _unrelated_ reasons, I think those process changes still reflect the user's latest thinking on how best to move Freight into that Stage, and for that reason alone, perhaps should not be disregarded.

imho, we should consider retrying a failed Promotion by just copying the Stage name and Freight name to a new Promotion and letting everything play out as usual.

I'd still like to see it implemented as a new API endpoint, even if reusing the process as it was defined previously defined isn't the impetus for it. As noted, the CLI would for sure use it as well.

**Comment by Marvin9** [comment]
UI part to re-run terminated promotion is available in 1.6 https://github.com/akuity/kargo/pull/4428

**Comment by krancour** [comment]
This works nicely. Reducing priority on the idea of a "retry" endpoint. It's not clear that we _need_ it right now.

**Comment by jessesuen** [comment]
Yep, the UI only feature might be good enough. Updating title since we have the UI feature.

**Comment by krancour** [comment]
I still feel it's worth doing. Just lower in priority at the moment.

But this also might be easy enough to be help-wanted (but probably too hard to be a good first issue).
